### PR TITLE
fix: fix for database insert failures

### PR DESCRIPTION
**Pull request by oderayi**

### CODE REVIEW

File: src/shared/setup.js [hunk 1]
```diff
@@ -47,6 +48,12 @@ async function connectDatabase () {
   await Db.connect(Config.DATABASE)
   const dbLoadedTables = Db._tables ? Db._tables.length : -1
   Logger.isDebugEnabled && Logger.debug(`DB.connect loaded '${dbLoadedTables}' tables!`)
+
+  // Connect central-ledger database as well
```
Comment (kalinkrustev) [inline src/shared/setup.js:52]: why is this needed?

Comment (oderayi) [reply src/shared/setup.js:52]: central-settlement uses models from central-ledger and those models need to have knex initialized. Otherwise, CL model calls will fail.

Comment (kalinkrustev) [reply src/shared/setup.js:52]: Ok, maybe put that detail in the comment.

File: src/models/settlementWindow/facade.js [hunk 2]
```diff
@@ -231,7 +260,9 @@ const Facade = {
...
                 .andWhere('m.settlementGranularityId', Enum.Settlements.SettlementGranularity.NET)
                 .groupBy('swc.settlementWindowContentId', 'pc.participantCurrencyId', 'tp.transferParticipantRoleTypeId', 'tp.ledgerEntryTypeId')
-                .select('swc.settlementWindowContentId', 'pc.participantCurrencyId', 'tp.transferParticipantRoleTypeId', 'tp.ledgerEntryTypeId',
+                .select('swc.settlementWindowContentId', 'pc.participantCurrencyId',
+                  knex.raw('CASE WHEN ?? = NULL THEN ? ELSE ?? END AS transferParticipantRoleTypeId', ['tp.transferParticipantRoleTypeId', 7, 'tp.transferParticipantRoleTypeId']), // @todo 7 = Enum.Accounts.TransferParticipantRoleType.COUNTER_PARTY_FSP requires upgrade of cs-shared
+                  knex.raw('CASE WHEN ?? = NULL THEN ? ELSE ?? END AS ledgerEntryTypeId', ['tp.ledgerEntryTypeId', Enum.Accounts.LedgerEntryType.PRINCIPLE_VALUE, 'tp.ledgerEntryTypeId']),
                   knex.raw('? AS ??', [Enum.Settlements.SettlementWindowState.CLOSED, 'settlementWindowStateId']),
                   knex.raw('? AS ??', [transactionTimestamp, 'createdDate']))
                 .sum('tp.amount AS amount')
```
Comment (vijayg10) [inline src/models/settlementWindow/facade.js:268]: Is the amount still from transferParticipant table? I think we can take it from ppc.change as the transferParticpant is left join and the value can be null

File: src/shared/setup.js (hunk 1 above)
Comment (oderayi) [reply src/shared/setup.js]: Added

File: src/models/settlementWindow/facade.js [hunk 3]
```diff
@@ -259,17 +255,44 @@ const Facade = {
...
-                .groupBy('swc.settlementWindowContentId', 'pc.participantCurrencyId', 'tp.transferParticipantRoleTypeId', 'tp.ledgerEntryTypeId')
+                .groupBy('swc.settlementWindowContentId', 'pc.participantCurrencyId', 'transferParticipantRoleTypeId', 'ledgerEntryTypeId')
                 .select('swc.settlementWindowContentId', 'pc.participantCurrencyId',
-                  knex.raw('CASE WHEN ?? = NULL THEN ? ELSE ?? END AS transferParticipantRoleTypeId', ['tp.transferParticipantRoleTypeId', 7, 'tp.transferParticipantRoleTypeId']), // @todo 7 = Enum.Accounts.TransferParticipantRoleType.COUNTER_PARTY_FSP requires upgrade of cs-shared
-                  knex.raw('CASE WHEN ?? = NULL THEN ? ELSE ?? END AS ledgerEntryTypeId', ['tp.ledgerEntryTypeId', Enum.Accounts.LedgerEntryType.PRINCIPLE_VALUE, 'tp.ledgerEntryTypeId']),
+                  knex.raw('CASE WHEN ppc.change > 0 THEN ? ELSE ?? END AS transferParticipantRoleTypeId', [ Enum.Accounts.TransferParticipantRoleType.PAYER_DFSP, Enum.Accounts.TransferParticipantRoleType.PAYEE_DFSP]),
+                  knex.raw('? AS ??', [Enum.Accounts.LedgerEntryType.PRINCIPLE_VALUE, 'ledgerEntryTypeId']),
```
Comment (vijayg10) [inline src/models/settlementWindow/facade.js:261]: @oderayi please observe the refactor I did here to avoid use of transferParticipant table and to accommodate FX

Comment (vijayg10) [reply src/models/settlementWindow/facade.js:261]: And also please not, we don't need to use other role types like initiatingFsp...etc. In settlements worl, we can just stick with PAYER_DFSP and PAYEE_DFSP. Otherwise we may need to change other settlement queries for the new role types.

Comment (PaulGregoryBaker) [reply src/models/settlementWindow/facade.js:261]: This change is cleaner.
LedgerEntryType can also be an Interchange_FEE. Not sure how this change would affect that.
I.e. Could any of the position changes be Interchange_FEE's? 
Here is where they are inserted https://github.com/mojaloop/central-settlement/blob/main/src/models/transferSettlement/facade.js#L40

File: src/models/settlementWindow/facade.js [hunk 4]
```diff
@@ -230,15 +255,44 @@ const Facade = {
...
+                .select('swc.settlementWindowContentId', 'pc.participantCurrencyId',
+                  knex.raw('CASE WHEN ppc.change > 0 THEN ? ELSE ?? END AS transferParticipantRoleTypeId', [ Enum.Accounts.TransferParticipantRoleType.PAYER_DFSP, Enum.Accounts.TransferParticipantRoleType.PAYEE_DFSP]),
+                  knex.raw('? AS ??', [Enum.Accounts.LedgerEntryType.PRINCIPLE_VALUE, 'ledgerEntryTypeId']),
+                  knex.raw('? AS ??', [Enum.Settlements.SettlementWindowState.CLOSED, 'settlementWindowStateId']),
+                  knex.raw('? AS ??', [transactionTimestamp, 'createdDate']))
+                .sum('ppc.change AS amount')
+            })
+            .transacting(trx)
+          await fxBuilder
+
```
Comment (kalinkrustev) [inline src/models/settlementWindow/facade.js:295]: I have some concerns about having a second query that inserts into settlementContentAggregation. I suspect this table might have some intention of aggregating amounts grouped over the fields settlementWindowContentId, participantCurrencyId, transferParticipantRoleTypeId and ledgerEntryTypeId. Is this a correct statement?
If it is correct, I do not see a unique constraint for these columns and maybe we must change the 2 queries to a single insert, using union and aggregation on top of it.

Comment (vijayg10) [reply src/models/settlementWindow/facade.js]: Done, thanks for spotting

File: src/models/settlementWindow/facade.js [hunk 5]
```diff
@@ -216,25 +228,62 @@ const Facade = {
...
-                .join('transferParticipant AS tp', 'tp.transferId', 'tf.transferId')
-                .join('participantCurrency AS pc', 'pc.participantCurrencyId', 'tp.participantCurrencyId')
+              this.from(function() {
+                this.select('ppc.*', 'tf.settlementWindowId')
+                  .from('transferFulfilment AS tf')
+                  .join('transferStateChange AS tsc', 'tsc.transferId', 'tf.transferId')
+                  .join('participantPositionChange AS ppc', 'ppc.transferStateChangeId', 'tsc.transferStateChangeId')
+                  .where('tf.settlementWindowId', settlementWindowId)
+                  .unionAll(function() {
+                    this.select('ppc.*', 'fxtf.settlementWindowId')
```
Comment (kalinkrustev) [inline src/models/settlementWindow/facade.js:264]: `select *` is a powerful way to break things when a new column appears in the table or columns are not ordered in the expected way, which can happen sometimes during the migrations. Please use explicit list of columns.

Comment (vijayg10) [reply src/models/settlementWindow/facade.js]: thanks, done

File: src/models/settlementWindow/facade.js [hunk 6]
```diff
@@ -50,6 +50,7 @@ const Facade = {
   },
 
   getTransfersCount: async function ({ settlementWindowId }) {
+    // Get fxTransferFulfilment count aswell?
```
Comment (kleyow) [inline src/models/settlementWindow/facade.js:53]: Can remove these comments from my earlier commit if they aren't needed

File: src/shared/setup.js [hunk 7]
```diff
@@ -47,6 +48,13 @@ async function connectDatabase () {
   await Db.connect(Config.DATABASE)
   const dbLoadedTables = Db._tables ? Db._tables.length : -1
   Logger.isDebugEnabled && Logger.debug(`DB.connect loaded '${dbLoadedTables}' tables!`)
+
+  // Initalize knex in central-ledger depedency
+  // This is required to be able to use the central-ledger models
```
Comment (kleyow) [inline src/shared/setup.js:53]: I don't see where these models are getting used?

**Review by vijayg10 — APPROVED** [review]
